src/sentry/api/endpoints/organization_stats_v2.py

```diff
@@ -0,0 +1,225 @@
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [organization.id], "project_id": project_ids},
+            orderby=["times_seen", "time"],
+        )
+        template = {
+            "statsErrors": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsTransactions": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsAttachments": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+        }
+        response = {project_id: deepcopy(template) for project_id in project_ids}
+
+        # group results by projectid>timestamp, using defaultdict to coalesce into format
+        for row in result:
+            nested_update(
+                response[row["project_id"]][CATEGORY_NAME_MAP[row["category"]]][row["time"]],
+                map_row_to_format(row),
+            )
+        # add project_ids with no results to dict
+        for project_id in project_ids:
+            if project_id not in response:
+                response[project_id] = deepcopy(template)
+
+        # zerofill response
+        response = {
+            project_id: {
+                category: zerofill(list(stats.values()), start, end, rollup, "time")
+                for category, stats in timeseries.items()
+            }
+            for project_id, timeseries in response.items()
+        }
+        return Response(response)
+
+
+class OrganizationProjectStatsDetails(ProjectEndpoint, OrganizationEndpoint):
+    def get(self, request, project):
+        start, end, rollup = get_date_range_rollup_from_params(request.GET, "1h", round_range=True)
+
+        # TODO: see if there's a better way to get the user's projects
+        project_list = []
+        team_list = Team.objects.get_for_user(organization=project.organization, user=request.user)
+        for team in team_list:
+            project_list.extend(Project.objects.get_for_user(team=team, user=request.user))
+        project_ids = list({p.id for p in project_list})
+
+        if project.id not in project_ids:
+            return Response(status=404)
+
+        result = outcomes.query(
+            start=start,
+            end=end,
+            rollup=rollup,
+            groupby=["category", "time", "outcome", "reason"],
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [project.organization.id], "project_id": [project.id]},
+            orderby=["time"],
+        )
+
+        response = {
+            "statsErrors": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsTransactions": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsAttachments": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+        }
+        for row in result:
+            nested_update(
+                response[CATEGORY_NAME_MAP[row["category"]]][row["time"]], map_row_to_format(row)
+            )
```

Comment:
I'm seeing a few readability issues here, which I think could be improved in part by introducing one or more classes to model the data that you're building to put into the response, and only converting it to a raw serializable value at the end.

If I'm getting this correctly, there are three layers of structure to model. The topmost is the data categories, the second is the structure of `DEFAULT_TS_VAL`, and the third is `{"quantity": 0, "times_seen": 0}`. If I were writing this from scratch and trying to maximize clarity, I'd probably write one class for each of those three.

One advantage is that the second class could just be a flat structure of five values ("accepted", "filtered", "overQuota", "spikeProtection", "other"), where you only worry about wrapping the last three into the `dropped` dict at the final serialization step.

If you're interested in undertaking this as a refactoring, I'll pair-program on it if you like.

Reply:
yes, good call on all of this. will give modeling this w/ classes a shot and making it more readable, will reach out for a pair-program if i'm feeling anything is ambiguous. thank you!